In [1]:
import pandas as pd
import numpy as np
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix
from sklearn import svm
from sklearn.model_selection import GridSearchCV
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score

In [2]:
df_raw = pd.read_csv('/Users/adityasingh/Downloads/Bank-Marketing-Dataset/train.csv', delimiter=';', encoding='UTF-8')
df_raw.head()
df_raw.shape

(45211, 17)

***EDA***

***Train Set***

In [3]:
df_raw.describe(include = 'all')

age          job  marital  education default        balance  \
count   45211.000000        45211    45211      45211   45211   45211.000000   
unique           NaN           12        3          4       2            NaN   
top              NaN  blue-collar  married  secondary      no            NaN   
freq             NaN         9732    27214      23202   44396            NaN   
mean       40.936210          NaN      NaN        NaN     NaN    1362.272058   
std        10.618762          NaN      NaN        NaN     NaN    3044.765829   
min        18.000000          NaN      NaN        NaN     NaN   -8019.000000   
25%        33.000000          NaN      NaN        NaN     NaN      72.000000   
50%        39.000000          NaN      NaN        NaN     NaN     448.000000   
75%        48.000000          NaN      NaN        NaN     NaN    1428.000000   
max        95.000000          NaN      NaN        NaN     NaN  102127.000000   

       housing   loan   contact           day  month      duration  \
count    45211  45211     45211  45211.000000  45211  45211.000000   
unique       2      2         3           NaN     12           NaN   
top        yes     no  cellular           NaN    may           NaN   
freq     25130  37967     29285           NaN  13766           NaN   
mean       NaN    NaN       NaN     15.806419    NaN    258.163080   
std        NaN    NaN       NaN      8.322476    NaN    257.527812   
min        NaN    NaN       NaN      1.000000    NaN      0.000000   
25%        NaN    NaN       NaN      8.000000    NaN    103.000000   
50%        NaN    NaN       NaN     16.000000    NaN    180.000000   
75%        NaN    NaN       NaN     21.000000    NaN    319.000000   
max        NaN    NaN       NaN     31.000000    NaN   4918.000000   

            campaign         pdays      previous poutcome      y  
count   45211.000000  45211.000000  45211.000000    45211  45211  
unique           NaN           NaN           NaN        4      2  
top              NaN           NaN           NaN  unknown     no  
freq             NaN           NaN           NaN    36959  39922  
mean        2.763841     40.197828      0.580323      NaN    NaN  
std         3.098021    100.128746      2.303441      NaN    NaN  
min         1.000000     -1.000000      0.000000      NaN    NaN  
25%         1.000000     -1.000000      0.000000      NaN    NaN  
50%         2.000000     -1.000000      0.000000      NaN    NaN  
75%         3.000000     -1.000000      0.000000      NaN    NaN  
max        63.000000    871.000000    275.000000      NaN    NaN

Results from EDA above:
Numeric Data:
1) Age range from 18 - 95, mean = 40, std = 10.6
2) Balance - 1362 mean, std = 3044
Binary Data:
1) Default history - Yes / No
2) Housing loan - Yes / No
3) Personal Loan - Yes / No
Categorical Data:
1) Job - 12 categories
2) Marital - 3 categories
3) Education - 4 categories

9) contact: contact communication type (categorical: "unknown","telephone","cellular")
10) day: last contact day of the month (numeric)
11) month: last contact month of year (categorical: "jan", "feb", "mar", …, "nov", "dec")
12) duration: last contact duration, in seconds (numeric)
13) campaign: number of contacts performed during this campaign and for this client (numeric, includes last contact)
14) pdays: number of days that passed by after the client was last contacted from a previous campaign (numeric, -1 means client was not previously contacted)
15) previous: number of contacts performed before this campaign and for this client (numeric)
16) poutcome: outcome of the previous marketing campaign (categorical: "unknown","other","failure","success")

In [4]:
df_raw.dtypes

age           int64
job          object
marital      object
education    object
default      object
balance       int64
housing      object
loan         object
contact      object
day           int64
month        object
duration      int64
campaign      int64
pdays         int64
previous      int64
poutcome     object
y            object
dtype: object

Binary variables to be mapped to 0 & 1 respectively

1) Default
2) Housing
3) Loan

In [5]:
df_raw['default'].replace({'no':0, 'yes':1},inplace=True)
df_raw['housing'].replace({'yes': 1, 'no': 0}, inplace=True)
df_raw['loan'].replace({'no':0, 'yes':1},inplace=True)

**Copying data frame to a new data frame before appying getdummies method for one-hot encoding**

In [6]:
df_work = df_raw.copy()
df_work.head()

age           job  marital  education  default  balance  housing  loan  \
0   58    management  married   tertiary        0     2143        1     0   
1   44    technician   single  secondary        0       29        1     0   
2   33  entrepreneur  married  secondary        0        2        1     1   
3   47   blue-collar  married    unknown        0     1506        1     0   
4   33       unknown   single    unknown        0        1        0     0   

   contact  day month  duration  campaign  pdays  previous poutcome   y  
0  unknown    5   may       261         1     -1         0  unknown  no  
1  unknown    5   may       151         1     -1         0  unknown  no  
2  unknown    5   may        76         1     -1         0  unknown  no  
3  unknown    5   may        92         1     -1         0  unknown  no  
4  unknown    5   may       198         1     -1         0  unknown  no

*Note that month column has <u>not</u> been converted to dummy values and is dropped for analysis*

***One Hot Encoding***

In [7]:
df_dummy = pd.get_dummies(df_work, columns=['job','marital','education','contact','poutcome'], drop_first=True)

***Standardization***

In [8]:
scaler = StandardScaler()

*Copying data to a new df again <u>after applying one-hot encoding</u> and before applying standardization*

In [9]:
# Applying standardization transformations to selected variables
df_standardized = df_dummy.copy()

In [10]:
# Renaming columns for easier comprehension
df_standardized.rename(columns={'campaign':'number of contacts'}, inplace=True)
df_standardized.rename(columns={'previous': 'previous # contacts'}, inplace=True)
df_standardized.rename(columns={'pdays':'#days gap b/w contact'}, inplace=True)

In [11]:
## Replacing the '-1' which meant not contacted to '0'
df_standardized['#days gap b/w contact'].replace({-1: 0}, inplace=True)

Initiate and populate the train variables i.e. x_train and y_train

In [12]:
columns_order = ['age',
 'default',
 'balance',
 'housing',
 'loan',
 'day',
 'month',
 'duration',
 'number of contacts',
 '#days gap b/w contact',
 'previous # contacts',
 'job_blue-collar',
 'job_entrepreneur',
 'job_housemaid',
 'job_management',
 'job_retired',
 'job_self-employed',
 'job_services',
 'job_student',
 'job_technician',
 'job_unemployed',
 'job_unknown',
 'marital_married',
 'marital_single',
 'education_secondary',
 'education_tertiary',
 'education_unknown',
 'contact_telephone',
 'contact_unknown',
 'poutcome_other',
 'poutcome_success',
 'poutcome_unknown',
'y']

In [13]:
df_standardized = df_standardized[columns_order]

In [14]:
x_train = df_standardized.iloc[:,0:-1]

In [15]:
x_train.drop('month', inplace=True, axis=1)
x_train.shape

(45211, 31)

In [16]:
scaler.fit(x_train)
scaler.transform(x_train)

array([[ 1.60696496, -0.13548989,  0.25641925, ..., -0.20597248,
        -0.185948  ,  0.47251925],
       [ 0.28852927, -0.13548989, -0.43789469, ..., -0.20597248,
        -0.185948  ,  0.47251925],
       [-0.74738448, -0.13548989, -0.44676247, ..., -0.20597248,
        -0.185948  ,  0.47251925],
       ...,
       [ 2.92540065, -0.13548989,  1.42959305, ..., -0.20597248,
         5.37784754, -2.11631591],
       [ 1.51279098, -0.13548989, -0.22802402, ..., -0.20597248,
        -0.185948  ,  0.47251925],
       [-0.37068857, -0.13548989,  0.52836436, ...,  4.85501757,
        -0.185948  , -2.11631591]])

In [17]:
y_train = df_standardized.iloc[:,-1]

In [18]:
y_train.replace({'no':0, 'yes':1}, inplace=True)

***Test Data***

In [19]:
df_test = pd.read_csv('/Users/adityasingh/Downloads/Bank-Marketing-Dataset/test.csv', encoding='UTF-8', delimiter=';')
df_test.head()

age          job  marital  education default  balance housing loan  \
0   30   unemployed  married    primary      no     1787      no   no   
1   33     services  married  secondary      no     4789     yes  yes   
2   35   management   single   tertiary      no     1350     yes   no   
3   30   management  married   tertiary      no     1476     yes  yes   
4   59  blue-collar  married  secondary      no        0     yes   no   

    contact  day month  duration  campaign  pdays  previous poutcome   y  
0  cellular   19   oct        79         1     -1         0  unknown  no  
1  cellular   11   may       220         1    339         4  failure  no  
2  cellular   16   apr       185         1    330         1  failure  no  
3   unknown    3   jun       199         4     -1         0  unknown  no  
4   unknown    5   may       226         1     -1         0  unknown  no

In [20]:
df_test.shape

(4521, 17)

In [21]:
def df_transform(df):
    df = pd.get_dummies(df, columns=['job','marital','education','contact','poutcome'], drop_first=True)
    df['previous'] = scaler.fit_transform(df[['previous']])
    df.rename(columns={'campaign':'number of contacts'}, inplace=True)
    df.rename(columns={'pdays': '#days gap b/w contact'}, inplace=True)
    df.rename(columns={'previous': 'previous # contacts'}, inplace=True) 
    df['default'].replace({'no':0, 'yes':1},inplace=True)
    df['housing'].replace({'yes': 1, 'no': 0}, inplace=True)
    df['loan'].replace({'no':0, 'yes':1},inplace=True)  
    df['#days gap b/w contact'].replace({-1: 0}, inplace=True)
    df = df[columns_order]
    df.replace({'no':0, 'yes':1}, inplace=True)
    return df

In [22]:
df_transform(df_test)

age  default  balance  housing  loan  day month  duration  \
0      30        0     1787        0     0   19   oct        79   
1      33        0     4789        1     1   11   may       220   
2      35        0     1350        1     0   16   apr       185   
3      30        0     1476        1     1    3   jun       199   
4      59        0        0        1     0    5   may       226   
...   ...      ...      ...      ...   ...  ...   ...       ...   
4516   33        0     -333        1     0   30   jul       329   
4517   57        1    -3313        1     1    9   may       153   
4518   57        0      295        0     0   19   aug       151   
4519   28        0     1137        0     0    6   feb       129   
4520   44        0     1136        1     1    3   apr       345   

      number of contacts  #days gap b/w contact  ...  marital_single  \
0                      1                      0  ...               0   
1                      1                    339  ...               0   
2                      1                    330  ...               1   
3                      4                      0  ...               0   
4                      1                      0  ...               0   
...                  ...                    ...  ...             ...   
4516                   5                      0  ...               0   
4517                   1                      0  ...               0   
4518                  11                      0  ...               0   
4519                   4                    211  ...               0   
4520                   2                    249  ...               1   

      education_secondary  education_tertiary  education_unknown  \
0                       0                   0                  0   
1                       1                   0                  0   
2                       0                   1                  0   
3                       0                   1                  0   
4                       1                   0                  0   
...                   ...                 ...                ...   
4516                    1                   0                  0   
4517                    0                   1                  0   
4518                    1                   0                  0   
4519                    1                   0                  0   
4520                    0                   1                  0   

      contact_telephone  contact_unknown  poutcome_other  poutcome_success  \
0                     0                0               0                 0   
1                     0                0               0                 0   
2                     0                0               0                 0   
3                     0                1               0                 0   
4                     0                1               0                 0   
...                 ...              ...             ...               ...   
4516                  0                0               0                 0   
4517                  0                1               0                 0   
4518                  0                0               0                 0   
4519                  0                0               1                 0   
4520                  0                0               1                 0   

      poutcome_unknown  y  
0                    1  0  
1                    0  0  
2                    0  0  
3                    1  0  
4                    1  0  
...                ... ..  
4516                 1  0  
4517                 1  0  
4518                 1  0  
4519                 0  0  
4520                 0  0  

[4521 rows x 33 columns]

In [23]:
df_test.head()
df_test.shape

(4521, 17)

In [24]:
x_test = df_test.iloc[:,0:-1]

In [25]:
x_test.drop('month', inplace=True, axis =1)

In [26]:
x_test.head()

age          job  marital  education default  balance housing loan  \
0   30   unemployed  married    primary      no     1787      no   no   
1   33     services  married  secondary      no     4789     yes  yes   
2   35   management   single   tertiary      no     1350     yes   no   
3   30   management  married   tertiary      no     1476     yes  yes   
4   59  blue-collar  married  secondary      no        0     yes   no   

    contact  day  duration  campaign  pdays  previous poutcome  
0  cellular   19        79         1     -1         0  unknown  
1  cellular   11       220         1    339         4  failure  
2  cellular   16       185         1    330         1  failure  
3   unknown    3       199         4     -1         0  unknown  
4   unknown    5       226         1     -1         0  unknown

In [28]:
set(x_test.columns) - set(x_train.columns)

{'campaign',
 'contact',
 'education',
 'job',
 'marital',
 'pdays',
 'poutcome',
 'previous'}

In [27]:
scaler.transform(x_test)
x_test.shape

ValueError: The feature names should match those that were passed during fit.
Feature names unseen at fit time:
- age
- balance
- campaign
- contact
- day
- ...


In [ ]:
y_test = df_test.iloc[:,-1]

In [ ]:
y_test.replace({'no':0, 'yes':1}, inplace=True)

In [ ]:
y_test.shape

In [ ]:
y_train.shape

# Logistic Regression

*Using GridSearchCV to tune hyperparameters for Logistic Regression*

In [ ]:
param_grid_LR = {
    'solver' : ['lbfgs', 'liblinear', 'newton-cg', 'newton-cholesky', 'sag', 'saga'],
    'C' : [0.1, 1, 10, 100]
}

In [ ]:
# LR = LogisticRegression(max_iter=1000)
# clf = GridSearchCV(LR, param_grid_LR)
# clf.fit(x_train, y_train)

In [ ]:
# clf.best_params_

In [ ]:
# Best parameters for Logistic Regression: C : 1, solver : 'liblinear' as per above test
LR = LogisticRegression(random_state = 0, max_iter = 1000, C=1, solver='liblinear').fit(x_train, y_train)

In [ ]:
yhat_lr = LR.predict(x_test)

Checking Model Accuracy For Logistic Regression

Confusion Matrix

In [ ]:
import seaborn as sns
cm_lr = confusion_matrix(y_test, yhat_lr)

In [ ]:
sns.heatmap(cm_lr/np.sum(cm_lr), annot=True, fmt='.2%', cmap='Blues')

In [ ]:
print (classification_report(y_test, yhat_lr))

# K Nearest Neighbours

In [ ]:
# Ks = 10
# mean_acc = np.zeros((Ks))

# for n in range(1,Ks):
    
#     #Train Model and Predict  
#     neigh = KNeighborsClassifier(n_neighbors = n).fit(x_train.values,y_train.values)
#     yhat=neigh.predict(x_test.values)
#     mean_acc[n] = accuracy_score(y_test, yhat)

# mean_acc

We will select k = 3 as this gives us the highest accuracy and again print classification report

In [ ]:
k = 3
Knn = KNeighborsClassifier(n_neighbors=k)
Knn.fit(x_train.values, y_train.values)
yhat_knn = Knn.predict(x_test.values)

In [ ]:
cm_knn = confusion_matrix(y_test, yhat_knn)
sns.heatmap(cm_knn/np.sum(cm_knn), annot=True, cmap='Blues', fmt='.2%')

In [ ]:
print (classification_report(y_test, yhat_knn))

We find that KNearestNeighbors is better at classifying the given dataset basis classification report accuracy

# Support Vector Machine

In [ ]:
SVM_df = svm.SVC()

In [ ]:
# ,'poly','sigmoid'
parameters_grid_SVC = {'kernel':('linear', 'rbf','poly','sigmoid'),
              'C' : [0.1, 1, 10],
              }

In [ ]:
# clf = GridSearchCV(SVM_df, parameters_grid_SVC, n_jobs=-1)
# clf.fit(x_train, y_train)

In [ ]:
# clf.best_params_

In [ ]:
# {'C': 0.1, 'kernel': 'linear'} as per clf.best_params_, GridSearchCV takes approx 40 minutes to fit x_train and y_train data
clf_svm = svm.SVC(kernel='linear', C=0.1,probability=True)
clf_svm.fit(x_train, y_train)

In [ ]:
yhat_svm = clf_svm.predict(x_test)

*Checking accuracy*

In [ ]:
# Confusion matrix
cm_svm = confusion_matrix(y_test, yhat_svm)
sns.heatmap(cm_svm / np.sum(cm_svm), cmap = 'Blues', annot = True, fmt='.2%')

In [ ]:
print(classification_report(y_test, yhat_svm))

# Naive Bayes

*Gaussian distribution*

In [ ]:
from sklearn.naive_bayes import GaussianNB
clf = GaussianNB().fit(x_train, y_train)
yhat = clf.predict(x_test)
print(classification_report(y_test, yhat))

*Bernoulli Distribution with binarize set to True*

In [ ]:
from sklearn.naive_bayes import BernoulliNB
clf_nb = BernoulliNB(binarize=True).fit(x_train, y_train)
yhat_nb = clf_nb.predict(x_test)
print(classification_report(y_test, yhat_nb))

In [ ]:
cm_nb = confusion_matrix(y_test,yhat)
sns.heatmap(cm_nb/np.sum(cm_nb), fmt = '.2%', cmap='Blues', annot=True)

# Decision Tree

In [ ]:
from sklearn.tree import DecisionTreeClassifier
clf_dt = DecisionTreeClassifier()
clf_dt.fit(x_train, y_train)
yhat_dt = clf_dt.predict(x_test)
print(classification_report(y_test, yhat_dt))

In [ ]:
cm_dt = confusion_matrix(y_test, yhat_dt)
sns.heatmap(cm_dt/np.sum(cm_dt), annot=True, fmt = '.2%', cmap='Blues')

*Cost Complexity Pruning*

In [ ]:
from sklearn import tree
from matplotlib import pyplot as plt
plt.figure(figsize=(20,20))
tree.plot_tree(clf_dt,filled = True)

In [ ]:
path = clf_dt.cost_complexity_pruning_path(x_train, y_train)
ccp_alphas, impurities = path.ccp_alphas, path.impurities

In [ ]:
ccp_alphas

In [ ]:
clfs = []
for ccp_alpha in ccp_alphas:
    clf = DecisionTreeClassifier(random_state=0, ccp_alpha=ccp_alpha)
    clf.fit(x_train, y_train)
    clfs.append(clf)

*Accuracy vs Alpha for training & testing sets*

In [ ]:
print (x_train.shape, x_test.shape)

In [ ]:
train_scores = [clf.score(x_train, y_train) for clf in clfs]
test_scores = [clf.score(x_test, y_test) for clf in clfs]
fig, ax = plt.subplots()
ax.set_xlabel("alpha")
ax.set_ylabel("accuracy")   
ax.set_title("Accuracy vs alpha for training and testing sets")
ax.plot(ccp_alphas, train_scores, marker='o', label="train",drawstyle="steps-post")
ax.plot(ccp_alphas, test_scores, marker='o', label="test",drawstyle="steps-post")
ax.legend()
plt.show()

# Random Forest

In [ ]:
from sklearn.ensemble import RandomForestClassifier
clf_rf = RandomForestClassifier(n_estimators=50, criterion='gini', n_jobs= -1, random_state=42, max_depth=5, min_samples_split=20, min_samples_leaf=10)
clf_rf.fit(x_train, y_train)
yhat_rf = clf.predict(x_test)
cm_rf = confusion_matrix(y_test, yhat_rf)
sns.heatmap(cm_rf/np.sum(cm_rf), annot=True, fmt='0.2%', cmap = 'Blues')

In [ ]:
print(classification_report(y_test, yhat_rf))

# Selecting the best classification algorithm

In [ ]:
from sklearn.metrics import roc_curve, auc, roc_auc_score
import matplotlib.pyplot as plt

*Calculating ROC curve for Logistic Regression*

In [ ]:
yhat_lr_proba = LR.predict_proba(x_test)[:,1]
print(yhat_lr_proba)

In [ ]:
fpr, tpr, thresholds = roc_curve(y_test, yhat_lr_proba)
AUC_LR = roc_auc_score(y_test, yhat_lr_proba)
plt.plot(fpr, tpr, linewidth = 4)
plt.xlabel('False Positive Rate')
plt.ylabel('True Positive Rate')
plt.title('ROC Curve for Logistic Regression')
plt.show()

In [ ]:
print (AUC_LR*100)

*Calculating ROC curve for SVM*

In [ ]:
yhat_svm_proba = clf_svm.predict_proba(x_test)[:,1]
fpr, tpr, thresholds = roc_curve(y_test, yhat_svm_proba)
AUC_SVM = roc_auc_score(y_test, yhat_svm_proba)
plt.plot(fpr, tpr, linewidth = 4)
plt.xlabel('False Positive Rate')
plt.ylabel('True Positive Rate')
plt.title('ROC Curve for KNN')
plt.show()

In [ ]:
print(AUC_SVM*100)

*Calculating ROC for Naive Bayes*

In [ ]:
yhat_nb_proba = clf_nb.predict_proba(x_test)[:,1]
fpr, tpr, thresholds = roc_curve(y_test, yhat_nb_proba)
AUC_NB = roc_auc_score(y_test, yhat_nb_proba)
plt.plot(fpr, tpr, linewidth = 4)
plt.xlabel('False Positive Rate')
plt.ylabel('True Positive Rate')
plt.title('ROC Curve for Naive Bayes')
plt.show()
print (AUC_NB*100)

*Calculating ROC for Decision Tree*

In [ ]:
yhat_dt_proba = clf_dt.predict_proba(x_test)[:,1]
fpr, tpr, thresholds = roc_curve(y_test, yhat_dt_proba)
AUC_DT = roc_auc_score(y_test, yhat_dt_proba)
plt.plot(fpr, tpr, linewidth = 4)
plt.xlabel('False Positive Rate')
plt.ylabel('True Positive Rate')
plt.title('ROC Curve for Decision Tree')
plt.show()
print (AUC_DT*100)

*Calculating ROC for Random Forest*

In [ ]:
yhat_rf_proba = clf_rf.predict_proba(x_test)[:,1]
fpr, tpr, thresholds = roc_curve(y_test, yhat_rf_proba)
AUC_RF = roc_auc_score(y_test, yhat_rf_proba)
plt.plot(fpr, tpr, linewidth = 4)
plt.xlabel('False Positive Rate')
plt.ylabel('True Positive Rate')
plt.title('ROC Curve for Random Forest')
plt.show()
print (AUC_RF*100)

Selected KNN basis best accuracy + greatest area under ROC